File: Design-Patterns-Using-Go/Flyweight/Flyweight2.go
package main

import "fmt"

// Flyweight interface
type Flyweight interface {
	Operation(extrinsicState string)
}

// ConcreteFlyweight type
type ConcreteFlyweight struct {
	intrinsicState string
}

func (cf *ConcreteFlyweight) Operation(extrinsicState string) {
	fmt.Println("Operation inside concrete flyweight.")
}

// FlyweightFactory type
type FlyweightFactory struct {
	flyweights map[string]Flyweight
}

func NewFlyweightFactory() *FlyweightFactory {
	return &FlyweightFactory{
		flyweights: make(map[string]Flyweight),
	}
}

func (ff *FlyweightFactory) GetFlyweight(intrinsicState string) Flyweight {
	if _, exists := ff.flyweights[intrinsicState]; !exists {
		ff.flyweights[intrinsicState] = &ConcreteFlyweight{intrinsicState: intrinsicState}
	}
	return ff.flyweights[intrinsicState]
}

// ClientClass type
type ClientClass struct {
	flyweight       Flyweight
	extrinsicState  string
}

func NewClientClass(factory *FlyweightFactory, intrinsicState, extrinsicState string) *ClientClass {
	return &ClientClass{
		flyweight:      factory.GetFlyweight(intrinsicState),
		extrinsicState: extrinsicState,
	}
}

func (cc *ClientClass) Operation() {
	fmt.Println("Operation inside context.")
	cc.flyweight.Operation(cc.extrinsicState)
}

// Client Code
func main() {
	factory := NewFlyweightFactory()
	c := NewClientClass(factory, "common", "separate1")
	c.Operation()

	c2 := NewClientClass(factory, "common", "separate2")
	c2.Operation()
}

/*
Operation inside context.
Operation inside concrete flyweight.
Operation inside context.
Operation inside concrete flyweight.
*/

File: Design-Patterns-Using-Kotlin/Flyweight/Flyweight2.kt
// Flyweight interface
interface Flyweight {
    fun operation(extrinsicState: String?)
}

// Concrete Flyweight class
class ConcreteFlyweight(private val intrinsicState: String) : Flyweight {
    override fun operation(extrinsicState: String?) {
        println("Operation inside concrete Flyweight: $this")
    }
}

// FlyweightFactory class
class FlyweightFactory {
    private val flyweights: MutableMap<String, Flyweight> = HashMap()

    fun getFlyweight(intrinsicState: String): Flyweight? {
        if (!flyweights.containsKey(intrinsicState)) {
            flyweights[intrinsicState] = ConcreteFlyweight(intrinsicState)
        }
        return flyweights[intrinsicState]
    }
}

// ClientClass
class ClientClass(factory: FlyweightFactory, intrinsicState: String, var extrinsicState: String) {
    var flyweight: Flyweight? = factory.getFlyweight(intrinsicState)

    fun operation() {
        println("Operation inside context: $this")
        flyweight!!.operation(extrinsicState)
    }
}

// Client Code
fun main() {
    val factory = FlyweightFactory()
    val c = ClientClass(factory, "common", "separate1")
    c.operation()
    val c2 = ClientClass(factory, "common", "separate2")
    c2.operation()
}

/*
Operation inside context: ClientClass@7d4991ad
Operation inside concrete Flyweight: ConcreteFlyweight@28d93b30
Operation inside context: ClientClass@1b6d3586
Operation inside concrete Flyweight: ConcreteFlyweight@28d93b30
*/

